### DISCUSSION TITLE
Comparison to telepresence

### DISCUSSION THREAD

**Discussion by tamasfe** [metalbear-co/mirrord#154] [General]
Hi, I've just found this project and I'm wondering how it relates to [telepresence](https://www.telepresence.io) that I use daily for development.

From a first glance of the documentation and feature plans, mirrord seems to target the same feature set as telepresence. If this is correct, I'd be interested in why this tool could be a better choice ~~(inb4 because it's written in Rust)~~ and what improvements one could expect by switching to it.

From what I see, the main difference for the existing features lies in the additional IDE extension. I'm also all for RIIR with the hope of improving past the plug and pray nature of telepresence.

**Comment by eyalb181** [comment]
Hi @tamasfe, thanks for taking an interest! You're right in pointing out that Telepresence is the closest alternative (that we know of) to mirrord. There are some key differences though:
- mirrord optimizes for concurrency and non-invasiveness, letting multiple developers work simultaneously on the same cloud environment without stepping on each other's toes or breaking the environment for everyone else. So in terms of functionality:
    - Letting you duplicate traffic rather than intercept it (we only support duplication at the moment, interception coming soon)
    - Fine-grained control over what actions happen remotely and what actions stay local. So you'd be able to filter by HTTP method or parameters, database table/collection and query format, file name/path, etc. (this isn't implemented yet)
    - Copy-on-Write for databases, so write commands would happen locally and read commands would happen remotely unless the entity being read was previously written locally, in which case it would be read locally as well (not implemented yet either).
- mirrord runs on the process level. Telepresence either runs on your local machine, changing its network behavior or in a container making development less smooth than running a local process. So we think we can achieve an overall smoother developer experience with mirrord.

We'd love to hear your thoughts - what do you use Telepresence for, and do you think mirrord could provide a good alternative given these differentiators? Are there any big gaps that we're overlooking?

**Reply by tamasfe** [reply]
Thanks, sounds great!

My use-cases would be the following:

1. intercept a service and route traffic to/from a locally running process instead
1. access services inside the cluster
1. copy environment variables from a secret/configmap or even running containers and set them locally

I would ideally want to do 2. and 3. without having to do 1. preferrably in an interactive shell environment.

Being able to do the following would be great:

```sh
# forward and resolve DNS for services and pods in the cluster
# and make accessing "whatever.cluster.local." just automagically work
mirrord exec --dns /bin/sh

# forward and resolve DNS and alias "foo-service" to "foo-service.foo.svc.cluster.local."
mirrord exec --namespace foo --dns /bin/sh

# set environment variables from containers in a pod, and then from a configmap
mirrord exec --namespace foo --env-from foo-pod --env-from foo-configmap /bin/sh

# intercept traffic from a pod, even inside the shell spawned from earlier commands,
# so services are already accessible and env vars are set
mirrod intercept --pod pod-name -p 8080 cargo run my-server
```

These are roughly the tasks I do regularly via clunky scripts and telepresence, I'm not sure how feasible they are to implement and how they align with the project goals.

**Reply by eyalb181** [reply]
That’s really interesting. So the use case we were thinking of was this: you’re developing version 1.1 of microservice X, and you’ve got version 1.0 running in the cluster - so if you want to see how v1.1 behaves in the cluster, you run it with mirrord. Point being - you have to pick some existing pod in the cluster to piggyback on.
Having said that, supporting your use case is definitely feasible, and it fits in really well with the general vision of mirrord (e.g. if I’m writing a Python service, I might want to test things in the cluster using a Python console).
So what you’re suggesting will definitely be supported with the caveat of having to run “from” a remote pod, and quite possibly be supported without that caveat :)

**Comment by aviramha** [comment]
Hey @tamasfe - we just released mirrord 3.0.1-alpha. We would love to get your feedback!

**Reply by tamasfe** [reply]
Awesome! I'll see if I can incorporate it into my dev process and will provide feedback/contributions.

**Reply by tamasfe** [reply]
I've tried it out, great stuff!

Mostly works as expected so far, there was nothing I couldn't find a workaround for.

I think currently mirrord does everything I used telepresence for (scoped to pods only, but I can live with it for now).

I've opened some issues, mostly DX nitpicks, feel free to assign those to me, I'd be happy to contribute.

**Reply by aviramha** [reply]
p.s regarding contributions - Let me know which you find exciting/cool for you to implement and I'll assign. I don't like picking issues for others as you'll know which you like the most :)
